vowpalwabbit/cb_explore_adf.cc

```diff
@@ -0,0 +1,484 @@
+
+    if(!is_learn || !data.learn_only) {      
+      if(data.tau) {
+	float prob = 1.0/(float)num_actions;
+	for(int i = 0;i < num_actions;i++) {
+	action_score a_s;
+	a_s.action = preds[i].action;
+	a_s.score = prob;
+	data.id_probs.push_back(a_s);
+	}
+	data.tau--;
+      }
+      else {
+	for(int i = 0;i < num_actions;i++) {
+	  action_score a_s;
+	  a_s.action = preds[i].action;
+	  a_s.score = 0.;
+	  data.id_probs.push_back(a_s);
+	}
+	data.id_probs[0].score = 1.0;
+      }    
+    }
+    
+    if (is_learn && data.known_cost.probability < 1)  
+      for(example* ec : examples)
+	base.learn(*ec);
+
+    for(size_t i = 0;i < num_actions;i++) {
+      examples[0]->pred.a_s[i].score = data.id_probs[i].score;
+      examples[0]->pred.a_s[i].action = data.id_probs[i].action;
+    }
+
+  }
+
+  template <bool is_learn>
+  void predict_or_learn_greedy(cb_explore_adf& data, base_learner& base, v_array<example*>& examples, bool isTest, bool shared)
+  { //Explore uniform random an epsilon fraction of the time.
+
+    v_array<action_score>& preds = examples[0]->pred.a_s;
+    data.id_probs.erase();
+
+    size_t num_actions = examples.size()-1;
+    if(shared)
+      num_actions--;
+    if(preds.size() != num_actions)
+      cerr<<"Received predictions of wrong size from CB base learner\n";
+
+    if(!is_learn || !data.learn_only) {
+      float prob = data.epsilon/(float)num_actions;
+      for(int i = 0;i < num_actions;i++) {
+	action_score a_s;
+	a_s.action = preds[i].action;
+	a_s.score = prob;
+	data.id_probs.push_back(a_s);
+      }
+      data.id_probs[0].score += (1-data.epsilon);
+    }
+        
+    
+    if (is_learn)
+      for(example* ec : examples)
+	base.learn(*ec);
+    
+    for(size_t i = 0;i < num_actions;i++) {
+      examples[0]->pred.a_s[i].score = data.id_probs[i].score;
+      examples[0]->pred.a_s[i].action = data.id_probs[i].action;
+    }
+
+  }
+
+  template <bool is_learn>
+  void predict_or_learn_bag(cb_explore_adf& data, base_learner& base, v_array<example*>& examples, bool isTest, bool shared)
+  { //Randomize over predictions from a base set of predictors
+
+    v_array<action_score>& preds = examples[0]->pred.a_s;
+    data.id_probs.erase();
+    
+    size_t num_actions = examples.size()-1;
+    if(shared)
+      num_actions--;
+    if(preds.size() != num_actions)
+      cerr<<"Received predictions of wrong size from CB base learner\n";
+
+    if(!is_learn || !data.learn_only) {
+      for(size_t i = 0;i < num_actions;i++) {
+	action_score a_s;
+	a_s.action = preds[i].action;
+	a_s.score = 0.;
+	data.id_probs.push_back(a_s);
+      }
+      float prob = 1.0/(float)data.bag_size;
+      data.id_probs[0].score += prob;
+      for(size_t i = 1;i < data.bag_size;i++) {
+	for(example* ec : examples)
+	  base.predict(*ec, i);
+	uint32_t chosen = preds[0].action;
+	for(size_t i = 0;i < num_actions;i++)
+	  if(data.id_probs[i].action == chosen)
+	    data.id_probs[i].score += prob;
+      }
```

Comment:
re: bag: Ah that makes sense, I missed the sorted order assumption. Also, the double counting is clearer upon closer look as i starts from 1. 

re: exploration library: Right, I think we're just talking about where the logic to compute prob should live (in exploration lib or vw). Let me try to find you offline to clarify.